### ISSUE TITLE
Keep block comment together with function assignment

### ISSUE THREAD

**Issue by emirotin** [jashkenas/coffeescript#3540]
Have a look at this example http://bit.ly/U51gmg:

```coffee
### 
File-level comment
###


### function F does this ###
f = ->

### function G does that ###
g = ->
```

2 functions are declared with some heredoc comments. I expect the comments to be kept with function bodies, but the 1st of them is separated with the list of hoisted variables.

Things can go even worse, for example: http://bit.ly/1noZeEN:

```coffee
### 
File-level comment
###


### function F does this ###
f = ->

### function G does that ###
g = ->

class X extends Y
```

It also has undesired side effect when comments are used for annotations read by 3rd party tools. In our case we use https://github.com/olov/ng-annotate which fails to interpret the 1st comment as it doesn’t see the function assignment after it.

Finally (a different issue, but related) CS inserts empty lines after heredocs, but inconsistently. When there’s no `var` statement it doesn’t: http://bit.ly/W4Cp3E. Which is actually a desired behavior — if I wish an empty line I can put it myself.
And in the 1st example CS doesn’t put empty line after the `F` comment, but does after the `G` comment.

**Comment by GeoffreyBooth** [comment]
This is fixed via #4572. Sorry it took awhile.

**Comment by emirotin** [comment]
Thank you.
